src/main/java/org/spongepowered/api/util/blockray/BlockRay.java

```diff
@@ -0,0 +1,420 @@
+            this.zCurrent = this.direction.getZ() * this.yPlaneT + this.position.getZ();
+            this.faceCurrent = this.yFace;
+            // Prepare next intersection
+            this.yPlaneNext += this.yPlaneIncrement;
+            this.yPlaneT = (this.yPlaneNext - this.position.getY()) / this.direction.getY();
+        } else {
+            // zPlaneT is smallest
+            this.xCurrent = this.direction.getX() * this.zPlaneT + this.position.getX();
+            this.yCurrent = this.direction.getY() * this.zPlaneT + this.position.getY();
+            this.zCurrent = this.zPlaneNext;
+            this.faceCurrent = this.zFace;
+            // Prepare next intersection
+            this.zPlaneNext += this.zPlaneIncrement;
+            this.zPlaneT = (this.zPlaneNext - this.position.getZ()) / this.direction.getZ();
+        }
+
+        this.blockCount++;
+
+        return this.lastHit = new BlockRayHit(this.extent, this.xCurrent, this.yCurrent, this.zCurrent, this.faceCurrent);
+    }
+
+    @Override
+    public void remove() {
+        throw new UnsupportedOperationException("Removal is not supported by this iterator");
+    }
+
+    // TODO: remove me once done
+    public static void test(CommandSource source) {
+        final Player player = (Player) source;
+        for (BlockRayHit hit : from(player).blockLimit(40)) {
+            if (!hit.getExtent().containsBlock(hit.getBlockPosition())) {
+                break;
+            }
+            //noinspection ConstantConditions
+            hit.getExtent().setBlockType(hit.getBlockPosition(), BlockTypes.DIAMOND_BLOCK);
+        }
+    }
+
+    /**
+     * Initializes a block ray builder, starting with the starting location.
+     *
+     * @param start The starting location
+     * @return A new block ray builder
+     * @see #BlockRay(BlockRayFilter, Location, Location)
+     * @see #BlockRay(BlockRayFilter, Location, Vector3d)
+     */
+    public static BlockRayBuilder from(Location start) {
+        Preconditions.checkArgument(start != null, "Start cannot be null");
+        return new BlockRayBuilder(start);
+    }
+
+    /**
+     * Initializes a block ray builder for the entity's eye.
+     * This sets both the starting point and direction.
+     *
+     * @param entity The entity
+     * @return A new block ray builder
+     */
+    public static BlockRayBuilder from(Entity entity) {
+        final Vector3d rotation = entity.getRotation();
+        final Vector3d direction = Quaterniond.fromAxesAnglesDeg(rotation.getY(), 360 - rotation.getX(), rotation.getZ()).getDirection();
+        // TODO: this needs to use the eye location
+        return from(entity.getLocation()).direction(direction);
+    }
+
+    /**
+     * A builder for block ray, which also implements {@link Iterable} which makes it
+     * useful for 'advanced for loops'. Use {@link #from(Location)} to get an instance.
+     */
+    public static class BlockRayBuilder implements Iterable<BlockRayHit> {
```

Comment:
In a lot of cases the plugin developer only wants to know the final ending location of the ray, so iterating it through seems a little long. I would suggest a `getEndingLocation()` method or `getEndingBlockRayHit`, but in these same cases location objects are more useful.